framework/core/instance.cpp

```diff
@@ -381,20 +381,54 @@ void Instance::query_gpus()
 
 PhysicalDevice &Instance::get_suitable_gpu()
 {
-	assert(!gpus.empty() && "No physical devices were found on the system.");
+	if (gpus.size() == 0)
+	{
+		throw std::runtime_error("No physical devices found.");
+	}
+
+	uint32_t count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU + 1];
+	memset(count_device_type, 0, sizeof(count_device_type));
+
+	for (uint32_t i = 0; i < gpus.size(); i++)
+	{
+		const auto deviceType = gpus[i]->get_properties().deviceType;
+		assert(deviceType <= VK_PHYSICAL_DEVICE_TYPE_CPU);
```

Comment:
You removed an assert at the top of the function. Seems a shame to add one back.

Reply:
They have different purpose.

The one at the top is still there, and turned into exception. Usage of the assert is wrong, as it will be removed in Release mode, but can still happen.

The one I added here, is impossible to happen, and tests internal API invariant.

Reply:
Not to nitpick, but technically it can happen in runtime without need for API changes in a weird and incompatible way.  All that's needed is for an additional device type enum to be defined.  While I'm unaware of a current need for a new device type, I wouldn't peg it as weird behavior if the enums are reordered or if a new device type enum were to be created.  I would agree if the device type enum listed a sentinal end that always comes with the guarantee that no enum would come after it, then it makes sense.